internal/batch/batch_future_test.go

```diff
@@ -0,0 +1,334 @@
+	s2.Chain(f)
+
+	wg := internal.NewWaitGroup(ctx)
+	wg.Add(1)
+	internal.GoNamed(ctx, "future-test", func(ctx internal.Context) {
+		defer wg.Done()
+		internal.Sleep(ctx, time.Second*10)
+		s.Set(1, nil)
+	})
+
+	err := f2.Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	err = f.Get(ctx, nil)
+	if err != nil {
+		return err
+	}
+
+	wg.Wait(ctx)
+	return err
+}
+
+func Test_Futures(t *testing.T) {
+	testSuite := &testsuite.WorkflowTestSuite{}
+	env := testSuite.NewTestWorkflowEnvironment()
+
+	env.RegisterWorkflow(futureTest)
+
+	env.ExecuteWorkflow(futureTest)
+}
+
+func batchWorkflowAssignWithSlice(ctx internal.Context) ([]int, error) {
+	totalSize := 5
+	concurrency := 2
+	factories := make([]func(ctx internal.Context) internal.Future, totalSize)
+	for i := 0; i < totalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+
+	var valuePtr []int
+	if err := batchFuture.Get(ctx, &valuePtr); err != nil {
+		return nil, err
+	}
+	return valuePtr, nil
+}
+
+func batchWorkflowAssignWithSliceOfPointers(ctx internal.Context) ([]int, error) {
+	totalSize := 5
+	concurrency := 2
+	factories := make([]func(ctx internal.Context) internal.Future, totalSize)
+	for i := 0; i < totalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+	batchFuture, err := NewBatchFuture(ctx, concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+	var valuePtr []*int
+	if err := batchFuture.Get(ctx, &valuePtr); err != nil {
+		return nil, err
+	}
+
+	var result []int
+	for _, v := range valuePtr {
+		result = append(result, *v)
+	}
+	return result, nil
+}
+
+func batchWorkflowAssignWithNil(ctx internal.Context) ([]int, error) {
+	totalSize := 5
+	concurrency := 2
+	factories := make([]func(ctx internal.Context) internal.Future, totalSize)
+	for i := 0; i < totalSize; i++ {
+		i := i
+		factories[i] = func(ctx internal.Context) internal.Future {
+			aCtx := internal.WithActivityOptions(ctx, internal.ActivityOptions{
+				ScheduleToStartTimeout: time.Second * 10,
+				StartToCloseTimeout:    time.Second * 10,
+			})
+			return internal.ExecuteActivity(aCtx, batchActivity, i)
+		}
+	}
+
+	batchFuture, err := NewBatchFuture(ctx, concurrency, factories)
+	if err != nil {
+		return nil, err
+	}
+
+	var valuePtr []int
+	if err := batchFuture.Get(ctx, nil); err != nil {
+		return nil, err
+	}
+	return valuePtr, nil
```

Comment:
_probably_ better to return an explicit `nil` here, just to avoid implying typed nils behave in a certain way (they should not, unless they have a custom marshaler)